streaming/src/main/scala/org/apache/spark/streaming/receiver/ReactiveReceiver.scala

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.receiver
+
+import org.reactivestreams._
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.annotation.DeveloperApi
+import scala.collection.mutable.ArrayBuffer
+import scala.util.{Try, Success, Failure}
+import java.lang.IllegalStateException
+
+/**
+ * An abstract class thet helps in implementing a Receiver compliant with the
+ * Reactive Streams API (http://www.reactive-streams.org/)
+ *
+ *
+ * Classes deriving from this API must be used with the 'reactive'
+ * congestion strategy setting. They can then communicate back-pressure through
+ * the reactiveCongestionStrategy below, and receive new elements through onNext.
+ *
+ * Implementation checklist:
+ *
+ * - Implementations should acquire a subscription in onStart by calling
+ *   Publisher#subscribe(this) on a reactive Publisher.
+ * - Implementations should override whenNext to configure custom behavior
+ *   upon receiving a new data element. They should call store in that method.
+ * - Implementations should override whenError to configure custom behavior
+ *   upon failure. They should call on Stop in that method.
+ * - Implementations should override whenComplete to configure custom behavior
+ *   upon completion. They should call onStop in that method.
+ *
+ * @see [[receiver.congestionStrategy]]
+ * @see [[org.reactivestreams.Subscriber]]
+ */
+@DeveloperApi
+abstract class ReactiveReceiver[T](storageLevel: StorageLevel)
+  extends Receiver[T](storageLevel) with Subscriber[T] { outer =>
+    private var subscription: Subscription = null
+    private var done: Boolean = false
+
+    private sealed trait SpecificationViolation {
+      val message: String
+      def printStackTrace(t: Option[Throwable] = None): Unit = t match {
+        case Some(e) => new IllegalStateException(message, e).printStackTrace(System.err)
+        case None => new IllegalStateException(message).printStackTrace(System.err)
+      }
+    }
+
+    private case class CancelException(s: Subscription) extends SpecificationViolation {
```

Comment:
Textual logs are harder to analyze with messages on multiple lines. Can we make this and the subsequent messages a single line?